coordinator/src/orderbook/mod.rs

```diff
@@ -30,120 +41,317 @@ pub mod websocket;
+        }
+
+        matched_orders
     }
-    pub fn remove(&mut self, order_id: Uuid) {
-        self.0.retain(|o| o.id != order_id)
+    /// rollback matched orders
+    pub fn rollback_matched_orders(&mut self, matched_orders: Vec<Order>) {
+        for order in matched_orders {
+            if let Some(order) = self.orders.iter_mut().find(|o| o.id == order.id) {
+                order.quantity += order.quantity;
+            } else {
+                self.add_order(order);
+            }
+        }
     }
-    pub fn sort(&mut self) {
-        self.0.sort_by(|a, b| {
+
+    /// Sorts the orders ascending if long and descending if short,
+    fn sort(&mut self, direction: commons::Direction) {
+        self.orders.sort_by(|a, b| {
             if a.price.cmp(&b.price) == Ordering::Equal {
                 return a.timestamp.cmp(&b.timestamp);
             }
-            // Ascending order.
-            a.price.cmp(&b.price)
-        });
+            match direction {
+                // Descending order.
+                commons::Direction::Short => b.price.cmp(&a.price),
+                // Ascending order.
+                commons::Direction::Long => a.price.cmp(&b.price),
+            }
+        })
     }
 }
 
 impl Orderbook {
-    pub fn new(pool: Pool<ConnectionManager<PgConnection>>) -> Self {
+    pub fn new(
+        pool: Pool<ConnectionManager<PgConnection>>,
+        order_matching_fee_rate: Decimal,
+    ) -> Self {
         Self {
             pool,
-            shorts: Shorts(vec![]),
-            longs: Longs(vec![]),
+            shorts: OrderbookSide::new(),
+            longs: OrderbookSide::new(),
+            order_matching_fee_rate,
         }
     }
 
+    /// Initializes the orderbook with non expired open limit orders.
     async fn initialize(&mut self) -> Result<()> {
         let all_orders = spawn_blocking({
             let mut conn = self.pool.clone().get()?;
             move || {
+                // TODO(holzeis): join with trades to get partially matched orders.
                 let all_orders =
                     orders::get_all_orders(&mut conn, OrderType::Limit, OrderState::Open, true)?;
                 anyhow::Ok(all_orders)
             }
         })
         .await??;
 
+        // TODO(holzeis): maybe consider batch adding all orders and sort afterwards instead of
+        // adding every order individually.
         for order in all_orders {
             match order.direction {
-                commons::Direction::Short => self.shorts.add(order.clone()),
-                commons::Direction::Long => self.longs.add(order.clone()),
+                commons::Direction::Short => self.shorts.add_order(order),
+                commons::Direction::Long => self.longs.add_order(order),
             }
         }
 
         Ok(())
     }
 
-    async fn add_order(
+    /// Adds a limit order to the orderbook.
+    async fn add_limit_order(&mut self, new_order: NewLimitOrder) -> Result<Message> {
```

Comment:
❓So we cannot match limit orders with limit orders yet?

Reply:
Yes not yet. I thought about addressing this when implementing limit orders for real 🙂